return {
	Hold = require(script.Objects.Hold),
	KeyOf = require(script.Objects.KeyOf),
	Compute = require(script.Objects.Compute),
	Absorb = require(script.Objects.Absorb),
	Listen = require(script.Objects.Listen),
	graph = require(script.Internal.graph),
	unpack = require(script.Utils.unpack),
	async = require(script.Internal.asyncInjector),
	wrapSource = require(script.Utils.wrapSource),
	sourceObject = require(script.Utils.sourceObject),
	equals = require(script.Utils.equals),
	clean = require(script.Utils.clean),
	alwaysTrue = require(script.Utils.alwaysTrue),
	createProcessor = require(script.Utils.createProcessor),
	destroyObject = require(script.Utils.destroyObject),
}
